File: HelloJNA/src/test/java/TestJNA.java
import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.Pointer;
import org.junit.jupiter.api.Test;

/**
 * @author MFine
 * @version 1.0
 * @date 2021/11/7 15:54
 **/
public class TestJNA {

    @Test
    void testVixHz_InitSDK() {
        CLibrary.INSTANCE.printf("fuck");
    }


    @Test
    void testJNA2() {
        Kernel32 INSTANCE = (Kernel32)
                Native.load("kernel32", Kernel32.class);
// Optional: wraps every call to the native library in a
// synchronized block, limiting native calls to one at a time
        Kernel32 SYNC_INSTANCE = (Kernel32)
                Native.synchronizedLibrary(INSTANCE);
        Kernel32.SYSTEMTIME time = new Kernel32.SYSTEMTIME();
        INSTANCE.GetSystemTime(time);

        System.out.println("Today's integer value is " + time.wDay);
    }


    @Test
    void testJNA3() {
        MYTestJNA.INSTANCE.sayHello();
        System.out.println("return from c++: " + MYTestJNA.INSTANCE.testReturnInt());
        int[] a = {1, 2, 3, 4};
        double[] d = new double[2];
         MYTestJNA.INSTANCE.returnArray(d, a, "lucky", 4);
        for (double lucky : d) {
            System.out.println(lucky);
        }
    }
}

File: jni-test/src/test-jna.cpp
//
// Created by MFine on 2021/11/7.
//


#include <iostream>

using namespace std;

void sayHello() {
    cout << "hello jna" << endl;
}

int testReturnInt() {
    return 1;
}

char const *returnArray(double d[], int a[], const char *str, int length) {
    for (int i = 0; i < length; i++) {
        cout << "from c++ value is d" << a[i] << endl;
    }
    cout << "from C++ str value: " << str << endl;
    while (*str) {
        cout << *str++ << endl;
    }
    d[0] = 1.0;
    d[1] = 4.0;
    char const *return_str = "hello from c++";
    return return_str;
}
